client/src/session_layer.rs

```diff
@@ -0,0 +1,679 @@
+        Box::pin(futures::future::ready(()))
+    }
+
+    fn on_request(
+        &self,
+        session_id: Vec<u8>,
+        request: proto::Request,
+        from: SocketAddr,
+    ) -> LocalBoxFuture<()> {
+        log::debug!("Received request packet from {}: {:?}", from, request);
+
+        let (request_id, kind) = match request {
+            proto::Request {
+                request_id,
+                kind: Some(kind),
+            } => (request_id, kind),
+            _ => return Box::pin(futures::future::ready(())),
+        };
+
+        match kind {
+            proto::request::Kind::Ping(request) => {
+                Box::pin(self.on_ping(session_id, request_id, from, request))
+            }
+            proto::request::Kind::Session(request) => {
+                Box::pin(self.dispatch_session(session_id, request_id, from, request))
+            }
+            _ => Box::pin(futures::future::ready(())),
+        }
+    }
+
+    fn on_forward(&self, forward: proto::Forward, from: SocketAddr) -> LocalBoxFuture<()> {
+        let myself = self.clone();
+        let fut = async move {
+            log::trace!(
+                "[{}] received forward packet ({} B) via {}",
+                myself.config.node_id,
+                forward.payload.len(),
+                from
+            );
+
+            let node = if forward.slot == 0 {
+                // Direct message from other Node.
+                match myself.state.read().await.nodes_addr.get(&from).cloned() {
+                    Some(id) => myself.registry.resolve_node(id).await?,
+                    None => {
+                        // In this case we can't establish session, because we don't have
+                        // neither NodeId nor SlotId.
+                        log::warn!(
+                            "Forward packet from unknown address: {}. Can't resolve.",
+                            from
+                        );
+                        return Ok(());
+                    }
+                }
+            } else {
+                // Messages forwarded through relay server or other relay Node.
+                match myself.registry.resolve_slot(forward.slot).await {
+                    Ok(node) => node,
+                    Err(_) => {
+                        log::debug!(
+                            "Forwarding from unknown Node (slot {}). Resolving..",
+                            forward.slot
+                        );
+
+                        // Try to establish session in case we can't find Node.
+                        let session = myself.server_session().await?;
+                        myself.resolve(session, forward.slot).await?
+                    }
+                }
+            };
+
+            if forward.is_reliable() {
+                myself.virtual_tcp.receive(node, forward.payload).await;
+            } else {
+                myself.dispatch_unreliable(node.id, forward).await;
```

Comment:
one calls `receive` and the other calls `dispatch`? seems naming is inconsistent

Reply:
`session_layer` and `virtual_tcp` have different API and semantic. I don't thing naming must be consistent, because they appear near each other

Reply:
IMHO all naming should be consistent in the whole code base

Reply:
For me it's not inconsistency